Add Prism Model

DaryaTereshchenko opened the pull request:
# What does this PR do?
Add a new translation model, `Prism` initially based on the [fairseq Transformer](https://github.com/facebookresearch/fairseq/tree/main/fairseq/models/transformer). 

The Prism model was first described in the paper [Automatic Machine Translation Evaluation in Many Languages via Zero-Shot Paraphrasing](https://aclanthology.org/2020.emnlp-main.8) (Thompson & Post, EMNLP 2020). Based on the code in this PR, I have converted the model to the HF model hub here: https://huggingface.co/dariast/prism

The original code can be found [here](https://github.com/thompsonb/prism/tree/master) and the original documentation is found [here](https://github.com/thompsonb/prism/blob/master/translation/README.md).

## Implementation notes
- The HF adaptation of the model is implemented utilizing existing [M2M100 model](https://github.com/huggingface/transformers/tree/main/src/transformers/models/m2m_100) architecture. The parts of the code that are identical to the HF M2M100 are cited accordingly. 
- The integration and performance tests are executed according to the guidelines and identical to the ones provided in the [M2M100 testing](https://github.com/huggingface/transformers/tree/main/tests/models/m2m_100) suite. 


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. 
@jvamvas

jvamvas commented on =0.26.0:
This can be deleted

jvamvas commented on src/transformers/models/prism/modeling_prism.py:
```diff
@@ -0,0 +1,1667 @@
+logger = logging.get_logger(__name__)
+
+_CONFIG_FOR_DOC = "PrismConfig"
+_CHECKPOINT_FOR_DOC = "dariast/prism"
```
Also link to this model repo in the PR message

jvamvas commented on src/transformers/models/prism/modeling_prism.py:
```diff
@@ -0,0 +1,1667 @@
+            encoder_attentions=outputs.encoder_attentions,
+        )
+
+    def prepare_inputs_for_generation(
```
I noticed that this is not in M2M100 – why was it necessary to override the parent method (https://github.com/huggingface/transformers/blob/c443d8d53685ae1d20b5f34e90c66312bd7b7a30/src/transformers/generation/utils.py#L348)? Did the parent method not work for Prism?

DaryaTereshchenko replied:
They recently moved this method to the GenerationMixin, when I started working on the Prism model using M2M100 the method prepare_inputs_for_generation() existed as well in the M2M100ForConditionalGeneration class. But, due to the recent changes it was moved to GenerationMixin. The initial implementation of M2M100 didn't inherit from GenerationMixin. I will try to accommodate my code and see how it works with the new  prepare_inputs_for_generation() method from GenerationMixin

jvamvas commented on src/transformers/models/prism/tokenization_prism.py:
Could it make sense to also have "# Copied from" strings, like in modeling_prism?

DaryaTereshchenko replied:
I can use the "# Copied from" strings only for the following methods because they remain unchanged ```prepare_seq2seq_batch, _build_translation_inputs, _switch_to_input_mode, _switch_to_target_mode, get_lang_token, get_lang_id```. Other methods were changed, and "# Copied from" strings can not be added because they rewrite the code according to the model from which the corresponding methods are copied. So, I want to make sure that no Prism-specific configurations will be rewritten

jvamvas commented on src/transformers/models/prism/configuration_prism.py:
Could it make sense to also have "# Copied from" strings, like in modeling_prism?

DaryaTereshchenko replied:
"# Copied from" added to  PrismOnnxConfig class

DaryaTereshchenko replied:
While running the ```make fix-copies``` it doesn't accept the '# Copied from' before ```PrismOnnxConfig``` therefore no comment added there

jvamvas commented on src/transformers/models/prism/configuration_prism.py:
```diff
@@ -0,0 +1,279 @@
+    Prism model according to the specified arguments, defining the model architecture. Instantiating a configuration
+    with the defaults will yield a similar configuration to that of the Prism architecture as described in the paper.
+
+    Args:
```
I noticed that the ordering of the arguments is a bit different from M2M100. If you make the ordering more similar, it would be easier to look at the code diff.

DaryaTereshchenko replied:
Fixed ✅

DaryaTereshchenko replied:
While running ```make quality``` provided in ```https://huggingface.co/docs/transformers/pr_checks```, the list of arguments in the config string changes, and some adaptations are needed. Therefore, the docstring for config that I have now is built according to the HuggingFace requirements.

jvamvas commented on tests/models/prism/test_modeling_prism.py:
```diff
@@ -0,0 +1,463 @@
+        parent,
+        batch_size=13,
+        seq_length=7,
+        is_training=False,
```
The original M2M100ModelTester has `is_training=True`, so it could make sense to be consistent here. But when I change this line, some tests fail. Do you know why?

```
FAILED tests/models/prism/test_modeling_prism.py::PrismModelTest::test_training - AttributeError: 'Seq2SeqModelOutput' object has no attribute 'loss'
FAILED tests/models/prism/test_modeling_prism.py::PrismModelTest::test_training_gradient_checkpointing - AttributeError: 'Seq2SeqModelOutput' object has no attribute 'loss'
FAILED tests/models/prism/test_modeling_prism.py::PrismModelTest::test_training_gradient_checkpointing_use_reentrant - AttributeError: 'Seq2SeqModelOutput' object has no attribute 'loss'
FAILED tests/models/prism/test_modeling_prism.py::PrismModelTest::test_training_gradient_checkpointing_use_reentrant_false - AttributeError: 'Seq2SeqModelOutput' object has no attribute 'loss'
```

DaryaTereshchenko replied:
The model was not registered in ```src/transformers/models/auto/configuration_auto.py``` and ```src/transformers/models/auto/modeling_auto.py```. These are the files responsible for recognizing the PrismModel as a base model that doesn't compute the loss and skipping tests that are not applicable.
The issue is resolved ✅

jvamvas commented on tests/models/prism/test_modeling_prism.py:
```diff
@@ -0,0 +1,463 @@
+
+    @cached_property
+    def default_tokenizer(self):
+        return PrismTokenizer.from_pretrained("dariast/prism")
```
Instead of hardcoding the model path, can you refer to the `checkpoint_name` variable of the class?
(same below)

DaryaTereshchenko replied:
Fixed ✅

jvamvas commented on tests/models/prism/test_modeling_prism.py:
```diff
@@ -0,0 +1,463 @@
+            forced_bos_token_id=tokenizer.get_lang_id("en"),
+        )
+
+        expected_en = [
```
My understanding is that both M2M100 and Prism prepend the target language ID to the translation, right? But I noticed that `expected_en` in the `M2M100ModelIntegrationTests` doesn't include the target language ID (https://github.com/huggingface/transformers/blob/e2ac16b28a0b8b900e136750309ca40c49d975c5/tests/models/m2m_100/test_modeling_m2m_100.py#L418), but this list does. Do you know why that is?
(same below)

DaryaTereshchenko replied:
The initial Prism model prepends the target language ID at the beginning of the encoded string for the decoder, and the output of the Fraiseq Prism model includes language ID during the conversion of generated tokens into strings. Moreover, M2M100 doesn't have language token IDs in the main vocabulary. Target language IDs are added to the ```special tokens``` list and processed differently to Prism tokenizer. In the case of the Prism model, target language IDs should be taken from the main vocabulary and should not be added to the ```special tokens```. Due to this difference, the ```convert_tokens_to_string``` that was identical to M2M100 before was changed so that no target language ID is visible in the final converted sequence. As of now, the Prism output doesn't have the target language ID, e.g., <en>, at the beginning of the converted sequence as it was before.

jvamvas replied:
Hi @DaryaTereshchenko, thanks for your response. I didn't look at your answer in detail yet, but what I can tell you is that the target language IDs should definitely be treated as special tokens. Maybe this will also resolve the remaining issues with the tokenizer, such as the `test_tokenizer_decode_ignores_language_codes` below.

Look at https://huggingface.co/facebook/m2m100_418M/blob/main/special_tokens_map.json to see how the language IDs can be configured as special tokens. (This should happen in the model repo, not in the code, because different model checkpoints might have different languages.)

jvamvas replied:
Closing this because `expected_en` now doesn't include the target language IDs, which was my original request in this conversation.

jvamvas commented on tests/models/prism/test_tokenization_prism.py:
```diff
@@ -0,0 +1,203 @@
+    SAMPLE_SP = get_tests_dir("fixtures/test_sentencepiece.model")
+
+
+if is_torch_available():
```
This statement doesn't seem necessary

DaryaTereshchenko replied:
Fixed ✅

jvamvas commented on tests/models/prism/test_tokenization_prism.py:
```diff
@@ -0,0 +1,203 @@
+class PrismTokenizerIntegrationTest(unittest.TestCase):
+    checkpoint_name = "dariast/prism"
+    src_text = ["Hi world.", "This is a Test.", "Some of my Best Friends are Linguists."]
+    tgt_text = ["<fr> Hé, monde!", "<fr> C'est un test.", "<fr> Certains de mes meilleurs amis sont linguistes."]
```
The `tgt_text` in `M2M100TokenizerIntegrationTest` doesn't contain the target language tokens: https://github.com/huggingface/transformers/blob/e2ac16b28a0b8b900e136750309ca40c49d975c5/tests/models/m2m_100/test_tokenization_m2m_100.py#L134
My understanding is that the tokenizers are very similar, so what explains this difference?

jvamvas replied:
`tgt_text` has been removed, but I noticed that the [`test_batch_fairseq_parity`](https://github.com/huggingface/transformers/blob/e2ac16b28a0b8b900e136750309ca40c49d975c5/tests/models/m2m_100/test_tokenization_m2m_100.py#L182) test from the `M2M100TokenizerIntegrationTest` is also missing. 
Is there a reason why this test method is not applicable to the Prism tokenizer?

jvamvas commented on tests/models/prism/test_tokenization_prism.py:
```diff
@@ -0,0 +1,203 @@
+    src_text = ["Hi world.", "This is a Test.", "Some of my Best Friends are Linguists."]
+    tgt_text = ["<fr> Hé, monde!", "<fr> C'est un test.", "<fr> Certains de mes meilleurs amis sont linguistes."]
+
+    expected_src_tokens = [37, 5050, 21, 1951, 13934, 33789, 7, 269, 11348, 983, 9393, 6, 2]
```
Consider using the `EN_CODE` constant you defined above

DaryaTereshchenko replied:
Fixed ✅

jvamvas commented on tests/models/prism/test_tokenization_prism.py:
```diff
@@ -0,0 +1,203 @@
+        cls.pad_token_id = 1
+        return cls
+
+    def test_language_codes(self):
```
The `M2M100TokenizerIntegrationTest` also has a test method `test_tokenizer_decode_ignores_language_codes`: https://github.com/huggingface/transformers/blob/e2ac16b28a0b8b900e136750309ca40c49d975c5/tests/models/m2m_100/test_tokenization_m2m_100.py#L166C9-L166C53.
Do you expect the same behavior with the Prism tokenizer? If so, it could make sense to add a similar method here

DaryaTereshchenko replied:
The test is added with a minor change. 
```self.assertIn(FR_CODE, self.tokenizer.all_special_ids)``` is not applicable to the Prism model because it doesn't include target language IDs in the list of special tokens.

jvamvas replied:
I assume with the fix you made, now it should be possible to add the line back.

DaryaTereshchenko replied:
Added the line ```self.assertIn(FR_CODE, self.tokenizer.all_special_ids)```  and tested, all tests pass.

jvamvas commented:
Hi @DaryaTereshchenko, thanks for the PR draft! This is really good work, I like how the code is already very clean!

- Moving the model to the `facebook` organization is not necessary here. It was necessary for my own PR (X-MOD), because that model was published by Meta. But the Prism paper is from Johns Hopkins University.
- For my own PR I also had to generate and edit a documentation page (https://github.com/huggingface/transformers/blob/7b4550f4998a2dd12e8d815279d69f6efa21eb82/docs/source/en/model_doc/xmod.mdx). Is that still part of the contribution guidelines? If so, this is also something this PR should include.
- I was able to successfully run the tests and to run the code example for translating a sentence. This is great! However, I noticed that the target language token is handled differently from M2M100. E.g., in your code example, translation is `"<fr> La vie comme une boîte de chocolat."`, and the code example from M2M100 doesn't include the `<fr>`. Could you please look into this? Thanks!

DaryaTereshchenko commented:
Hello @jvamvas 
Thank you for the comments! I tried to fix and accommodate as much as I could.

## Main fixes: 
1. "# Copied from" strings are tricky to use because they overwrite existing code and can be applied only to those classes and methods that are entirely identical to any existing methods or objects. As many methods of my tokenizer were adjusted and some logic was changed compared to M2M100, I kept it predominantly without "# Copied from" strings to avoid code overwrite.
2. Due to a different implementation of the convert_tokens_to_string method, the final output between M2M100 and Prism was different. But I fixed that. 
3. prepare_inputs_for_generation() was deleted because of inheritance from the GenerationMixin class.
4. All tests are executed without errors. An additional test for the tokenizer was added.
5. The model was added to the corresponding files necessary for correct mapping and documentation.
6. The documentation was built and tested.
7. The [model card](https://huggingface.co/dariast/prism) is also fixed.

Thank you!

jvamvas commented on docs/source/en/model_doc/prism.md:
```diff
@@ -0,0 +1,110 @@
+
+## Overview
+
+The `Prism` model, a state-of-the-art multilingual neural machine translation (NMT) system developed for translation. The model supports translation across 39 languages, leveraging a zero-shot paraphrasing approach that does not require human judgments for training.
```
Proposed rephrasing of this paragraph: "The Prism model, a multilingual neural machine translation (NMT) system."

Delete "state-of-the-art" ;)
The second sentence sounds a bit like: Prism leverages zero-shot paraphrasing for translation. Rather, the idea is that translation/paraphrasing can be leveraged for MT evaluation.

DaryaTereshchenko replied:
✅Fixed

jvamvas commented on docs/source/en/model_doc/prism.md:
```diff
@@ -0,0 +1,110 @@
+
+The `Prism` model, a state-of-the-art multilingual neural machine translation (NMT) system developed for translation. The model supports translation across 39 languages, leveraging a zero-shot paraphrasing approach that does not require human judgments for training.
+
+The `Prism` model was designed to be a lexically/syntactically unbiased paraphraser. The core idea is to treat paraphrasing as a zero-shot translation task, which allows the model to cover a wide range of languages effectively.
```
I don't think this second paragraph is necessary, because more context would be needed for readers to understand what this means.

DaryaTereshchenko replied:
✅Fixed

jvamvas commented on docs/source/en/model_doc/prism.md:
````diff
@@ -0,0 +1,110 @@
+# => 'Life is like a box of chocolate.'
+```
+
+## Languages Covered
````
Note that the list of supported languages is specific to a model checkpoint, not to the architecture itself. So I think the language list belongs into the Model Card, but not the documentation of the architecture.

DaryaTereshchenko replied:
✅Fixed

DaryaTereshchenko commented:
Hello @jvamvas 

Thank you for the comments! I have adjusted the ```prism.md```. 

But I'm afraid, I don't fully understand why my tokenizer should have the same [special_tokens_map](https://huggingface.co/facebook/m2m100_418M/blob/main/special_tokens_map.json) as M2M100. 

In the PrismTokenizer, language tokens (e.g., \<en\>, \<fr\>) are integrated directly into the main vocabulary. This means they are treated as regular tokens and are assigned IDs that fall within the primary range of the tokenizer's vocabulary.

The M2M100Tokenizer treats language tokens (e.g., \_\_en\_\_, \_\_fr\_\_) as additional special tokens. They are assigned unique IDs that are beyond the initial size of the main vocabulary (encoder_size + i), separating them from regular tokens.

While executing the M2M100 tokenizer code the special tokens are added to the dictionary automatically, which means that I have to adjust the Prism implementation. However, the part of handling special tokens is identical in [Prism and M2M100](https://github.com/jvamvas/transformers/blob/ca541bd4f4d932f486a4116deba833b4ffaebd15/src/transformers/models/m2m_100/tokenization_m2m_100.py#L134C1-L138C56), but compared to M2M100, the Prism language IDs are explicit regular tokens that should be in the main vocabulary: *"Note that the language tags are in the model vocabulary and should not be split using sentencepiece:"* [source](https://github.com/thompsonb/prism/blob/master/translation/README.md). 

Therefore, they are not added to the special_tokens_map.json. If I change the code and add them, then the model behaves differently, and many tests fail. 

I would appreciate your help with this issue very much because I don't think I understand how to resolve this problem.

Thank you!

jvamvas commented:
Hi @DaryaTereshchenko,
Thanks for your response.
I think there may be a few points that could use some clarification. For instance, there isn't such as thing as a "primary range" of a vocabulary. The size of the vocabulary is always equal to the height of the embedding matrix, regardless of whether we're dealing with special or regular tokens.

The only difference between special tokens and regular tokens is whether the tokenizer gives them special treatment during pre- or post-processing (like adding them before applying BPE, removing them after detokenization, or keeping them from merging with other tokens).

I suggest that we slow this down a bit. Could you try investing at least 4h into resolving the tokenizer / special tokens issue? Thanks.

DaryaTereshchenko commented:
Hello @jvamvas,

I took some time to look closely at how the additional_special_tokens list is constructed. I have found a way to add language tags to the list with special tokens correctly and tested them on existing tests for the tokenizer. The tests pass, and the modelling part works as expected. There are not many changes added, but it took some time to understand the logic. 

I have re-run the tests on the tokenizer and modelling files, and all tests pass. Please let me know if you encounter any failure, and I will work on it. Also, let me know what other fixes and adjustments to incorporate. 

Thank you for the help!

jvamvas commented on src/transformers/models/prism/tokenization_prism.py:
```diff
@@ -0,0 +1,426 @@
+        current_sub_tokens = []
+        out_string = ""
+        for token in tokens:
+            # Skip language tokens during decoding
+            if token in self.lang_code_to_token.values():
+                continue
```
I think that if the language tokens are correctly designated as special tokens, this "hack" will become unnecessary.

DaryaTereshchenko replied:
Fixed ✅

jvamvas reviewed (changes requested):
Hi @DaryaTereshchenko, thanks for your commit and sorry for the delayed review.

It definitely looks like a step in the right direction! For example, I see that the special_tokens_map.json of your model now lists the language IDs a special tokens, which is great.

However, there are still a few differences between M2M100 and Prism with regard to how the language IDs are handled:

```pycon
>>> from transformers import M2M100Tokenizer, PrismTokenizer

>>> m2m100_tokenizer = M2M100Tokenizer.from_pretrained("facebook/m2m100_418M")
>>> prism_tokenizer = PrismTokenizer.from_pretrained("dariast/prism")

>>> m2m100_tokenizer.decode(m2m100_tokenizer("Hello")["input_ids"])
'__en__ Hello</s>'
>>> prism_tokenizer.decode(prism_tokenizer("Hello")["input_ids"])
'Hello</s>'

>>> m2m100_tokenizer.all_special_tokens
['<s>', '</s>', '<unk>', '<pad>', '__af__', '__am__', '__ar__', '__ast__', '__az__', '__ba__', '__be__', '__bg__', '__bn__', '__br__', '__bs__', '__ca__', '__ceb__', '__cs__', '__cy__', '__da__', '__de__', '__el__', '__en__', '__es__', '__et__', '__fa__', '__ff__', '__fi__', '__fr__', '__fy__', '__ga__', '__gd__', '__gl__', '__gu__', '__ha__', '__he__', '__hi__', '__hr__', '__ht__', '__hu__', '__hy__', '__id__', '__ig__', '__ilo__', '__is__', '__it__', '__ja__', '__jv__', '__ka__', '__kk__', '__km__', '__kn__', '__ko__', '__lb__', '__lg__', '__ln__', '__lo__', '__lt__', '__lv__', '__mg__', '__mk__', '__ml__', '__mn__', '__mr__', '__ms__', '__my__', '__ne__', '__nl__', '__no__', '__ns__', '__oc__', '__or__', '__pa__', '__pl__', '__ps__', '__pt__', '__ro__', '__ru__', '__sd__', '__si__', '__sk__', '__sl__', '__so__', '__sq__', '__sr__', '__ss__', '__su__', '__sv__', '__sw__', '__ta__', '__th__', '__tl__', '__tn__', '__tr__', '__uk__', '__ur__', '__uz__', '__vi__', '__wo__', '__xh__', '__yi__', '__yo__', '__zh__', '__zu__']
>>> prism_tokenizer.all_special_tokens
['<s>', '</s>', '<unk>', '<pad>']
```

I think that as soon as these discrepancies are resolved, this will also automatically clear up the other issues that are currently open in the code review.

DaryaTereshchenko commented:
Hello @jvamvas,

I fixed the identified issues and checked the special tokens to be processed correctly. 
I mainly deleted the identified lines that I used to skip language tokens during decoding. I added the ```test_batch_fairseq_parity```; some numbers in assertions differ from M2M100 as I used different source texts. Also, as Prism requires the src_token ID, it is added in the ```vocab = ["</s>", "<unk>", "▁This", "▁is", "▁a", "▁t", "est", "\u0120", "<pad>", "<en>"]```. Therefore, in the test ```test_get_vocab```, the ```<s>``` has a different token ID than in the M2M100 model. But I checked, and in the M2M100 tests suite, they specify that the vocabulary length can differ, so I assume mine is correct. However, it should probably be double-checked, so I would appreciate your thoughts on this question. 

Thank you!

jvamvas commented:
@DaryaTereshchenko Thanks for the changes, looks good!
There is currently one unresolved comment in the review. After that, the PR is ready for Hugging Face :)

DaryaTereshchenko commented:
Hello, @jvamvas 
I added the missing line and tested both the modelling and tokenization parts. All tests pass.  ✅

Thank you!

jvamvas commented:
@DaryaTereshchenko Great! You may create the PR to HF now :)
